File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        count, ptr = 0, head
        while ptr:
            count += 1
            ptr = ptr.next
        
        index = count - n
        if index == 0:
            head = head.next
        else:
            ptr = head
            while index > 1:
                ptr = ptr.next
                index -= 1
            ptr.next = ptr.next.next
        return head

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: *mut ListNode,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: std::ptr::null_mut(), val }
//     }
// }

impl Solution {
    pub fn has_cycle(head: *mut ListNode) -> bool {
        if head.is_null() { return false; }
        unsafe {
            let mut slow = head;
            let mut fast = head;

            loop {
                slow = (*slow).next;
                fast = (*fast).next;
                if fast.is_null() { break; }
                fast = (*fast).next;
                if fast.is_null() { break; }

                if slow == fast {
                    return true;
                }
            }
            false
        }
    }
}
